lib/DependencyProvider.js

```diff
@@ -0,0 +1,317 @@
+// @flow
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const request = require('sync-request').default;
```

Comment:
TODO(lydell): Audit if we want this dependency. It seems pretty heavy (2MB): https://packagephobia.com/result?p=sync-request. @mpizenberg what does it do?

Reply:
So sync-request is here because it's not possible yet to have async callback functions in the pubgrub solver (Follow this https://github.com/pubgrub-rs/pubgrub/issues/110). So for now, both `fetchElmJson` and `listAvailableVersions` must be sync functions. I didn't know how to make sync http requests in Node so I found this package that enable sync http requests. If there is another way to make sync http requests don't hesitate to change this.

Reply:
I think I’d experiment replacing it with this approach (similar to how I do requests in `elm-tooling`):

1. Try to spawn `curl` (exists on macOS, Windows, some Linux) using [child_process.spawnSync](https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options)
2. If `curl` doesn’t exist (ENOENT), try to spawn `wget` (exists on some Linux) using using [child_process.spawnSync](https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options)
3. If none of them exist, spawn `node fallback-http.js` using [child_process.spawnSync](https://nodejs.org/api/child_process.html#child_processspawnsynccommand-args-options)

A benefit of using `curl` and `wget` is that people can then configure those to use proxies etc. And it should be much more lightweight.

Reply:
Your approach is better I think indeed. I don't plan on coding much until next week. If there is an easy way to reuse the elm-tooling approach don't hesitate to change it if you have the time. I can give it a try next week otherwise.

Reply:
Note for my future self, look at how we can make a sync request using a worker thread and Atomics: https://whistlr.info/2021/block-nodejs-main-thread/

Reply:
Not sure if that’s easier/better than what I suggested, but feel free to try!

Reply:
I've figured out a way to do this that seems to work base on this minimal example: https://github.com/mpizenberg/sync-get

However, when I do this inside node-test-runner, it's hanging. I haven't figured out yet why the request doesn't terminate and notify the main thread. I'll dig a little bit.

Reply:
I've pushed to the branch in case someone else wants to help diagnose the hanging issue.

Reply:
Alright I found the issue which turned out to be a very stupid issue of module paths. I had to replace `new Worker("./SyncGetWorker.js"` by `workerPath = path.resolve(__dirname, "SyncGetWorker.js");` and `new Worker(workerPath`

So good news, the dependency was removed, and by the way, this `SyncGetWorker` pattern might interest you @lydell (cf https://github.com/mpizenberg/sync-get).

What do you think?

PS: also fixed a bug due to forgetting to `JSON.parse` the get request for `elm.json` files to the package server.